// bottom-sheet-vue3/dist/index.mjs
import { ref as p, inject as L, withModifiers as M, defineComponent as P, onUnmounted as C, onMounted as O, createVNode as c, watch as A, Teleport as R, Transition as q, mergeProps as $ } from "vue";
const N = "data-bottom-sheet-stop", w = {
  type: Boolean,
  default: !1
}, B = {
  threshold: {
    type: Number,
    default: 100
  },
  onlyHeaderSwipe: w,
  noStretch: w,
  noClickOutside: w,
  noHeader: w
}, T = Symbol("BottomSheet");
function I() {
  const t = p(-1);
  let i = 0;
  const l = {
    activeSheet: t,
    id: () => ++i
  };
  return {
    install(S) {
      S.provide(T, l);
    }
  };
}
function Y() {
  return L(T);
}
function _(t, i) {
  return M(t, i);
}
const j = P({
  name: "SheetRenderer",
  props: {
    ...B,
    id: {
      type: Number,
      required: !0
    }
  },
  emits: {
    updateElement: null,
    close: null
  },
  setup(t, {
    emit: i,
    slots: l
  }) {
    const S = Y(), n = p(null);
    C(() => i("updateElement", null));
    let f = !0;
    const g = _(() => {
      t.noClickOutside || !f || i("close");
    }, ["self"]), d = p(!1), r = p(0), s = p(0);
    let o = !1, u = 0;
    function h() {
      n.value && (u = n.value.getBoundingClientRect().height);
    }
    function y() {
      var e, a;
      (e = n.value) == null || e.style.removeProperty("height"), (a = n.value) == null || a.style.removeProperty("max-height"), u = 0, d.value = !1, r.value = 0, s.value = 0;
    }
    function v(e) {
      const a = n.value;
      if (!a || !o)
        return;
      let m;
      if ("touches" in e ? (m = e.touches[0].clientY, e.preventDefault()) : m = e.clientY, m === s.value) {
        d.value = !1;
        return;
      }
      d.value = !0, r.value = m - s.value, f = !1, r.value < 0 && !t.noStretch ? (a.style.setProperty("height", `${u + Math.abs(r.value)}px`), a.style.setProperty("max-height", "none")) : (a.style.removeProperty("height"), a.style.removeProperty("max-height"));
    }
    function E(e) {
      e.target instanceof HTMLElement && e.target.closest(`[${N}]`) || ("touches" in e ? s.value = e.touches[0].clientY : s.value = e.clientY, o = !0);
    }
    function b() {
      const e = n.value;
      !o || !e || (r.value >= t.threshold ? i("close") : setTimeout(() => {
        f = !0, e.animate([{
          height: e.style.height
        }, {
          height: `${u}px`
        }], {
          duration: 200,
          easing: "ease"
        }).addEventListener("finish", () => {
          e.style.removeProperty("height"), e.style.removeProperty("max-height");
        });
      }), d.value = !1, s.value = 0, r.value = 0, o = !1);
    }
    function k(e) {
      t.onlyHeaderSwipe && E(e);
    }
    function H(e) {
      t.onlyHeaderSwipe || E(e);
    }
    const x = [["resize", y], ["resize", h], ["mousemove", v], ["mouseup", b], ["touchend", b], ["touchcancel", b], ["touchmove", v, {
      passive: !1
    }]];
    return O(() => {
      h();
      for (const [e, a, m] of x)
        window.addEventListener(e, a, m);
    }), C(() => {
      for (const [e, a] of x)
        window.removeEventListener(e, a);
    }), () => {
      var e;
      return c("div", {
        class: "bottom-sheet-backdrop",
        "data-sheet-active": t.id === S.activeSheet.value ? "" : null,
        draggable: !1,
        onClick: g,
        onMousedown: H,
        onTouchstart: H
      }, [c("div", {
        ref: n,
        class: "bottom-sheet",
        "data-swiping": d.value ? "" : null,
        style: {
          transform: `translateY(${Math.max(r.value, 0)}px)`
        }
      }, [t.noHeader ? null : c("div", {
        onMousedown: k,
        onTouchstart: k,
        class: "bottom-sheet-header"
      }, [l.header ? l.header() : c("div", {
        class: "bottom-sheet-header-bar"
      }, null)]), c("div", {
        class: "bottom-sheet-body"
      }, [(e = l.default) == null ? void 0 : e.call(l)])])]);
    };
  }
}), D = P({
  name: "Sheet",
  inheritAttrs: !1,
  props: {
    ...B,
    visible: {
      type: Boolean,
      required: !0
    }
  },
  emits: {
    "update:visible": null
  },
  setup(t, {
    attrs: i,
    slots: l,
    emit: S
  }) {
    const n = Y(), f = n.id();
    let g = n.activeSheet.value;
    A(() => t.visible, (o) => {
      o ? (g = n.activeSheet.value, n.activeSheet.value = f) : n.activeSheet.value = g;
    }, {
      immediate: !0
    });
    function d() {
      S("update:visible", !1);
    }
    function r(o, u) {
      const h = o.querySelector(".bottom-sheet"), y = window.getComputedStyle(o).backgroundColor;
      o.animate([{
        backgroundColor: "rgba(0, 0, 0, 0)"
      }, {
        backgroundColor: y
      }], {
        duration: 300,
        easing: "ease"
      });
      const v = h.animate([{
        transform: "translateY(100%)"
      }, {
        transform: "translateY(0%)"
      }], {
        duration: 300,
        easing: "ease"
      });
      v.onfinish = u;
    }
    function s(o, u) {
      const h = o.querySelector(".bottom-sheet"), y = window.getComputedStyle(o).backgroundColor;
      o.animate([{
        backgroundColor: y
      }, {
        backgroundColor: "rgba(0, 0, 0, 0)"
      }], {
        duration: 300,
        easing: "ease"
      });
      const v = h.animate([{
        transform: h.style.transform
      }, {
        transform: "translateY(100%)"
      }], {
        duration: 300,
        easing: "ease"
      });
      v.onfinish = u;
    }
    return C(() => {
      n.activeSheet.value = g;
    }), () => c(R, {
      to: "body"
    }, {
      default: () => [c(q, {
        css: !1,
        onEnter: r,
        onLeave: s
      }, {
        default: () => [t.visible && c(j, $({
          id: f,
          onClose: d
        }, i, {
          noClickOutside: t.noClickOutside,
          noHeader: t.noHeader,
          noStretch: t.noStretch,
          onlyHeaderSwipe: t.onlyHeaderSwipe,
          threshold: t.threshold
        }), l)]
      })]
    });
  }
});
export {
  D as Sheet,
  I as createBottomSheet
};
